File: 142. Linked List Cycle II.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode findConn(ListNode curr){
        HashMap<ListNode, Integer> hm = new HashMap<ListNode, Integer>();
        int ind = 0;
        while(!hm.containsKey(curr)){
            hm.put(curr, ind);
            ind++;
            curr = curr.next;
        }
        return curr;
    }
    public ListNode detectCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if(slow == fast) return findConn(head);
        }

        return null;
    }
}

File: 142. Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 */

class Solution {
public:
    ListNode* findConn(ListNode* curr) {
        std::unordered_map<ListNode*, int> hm;
        int ind = 0;
        while (hm.find(curr) == hm.end()) {
            hm[curr] = ind;
            ind++;
            curr = curr->next;
        }
        return curr;
    }

    ListNode* detectCycle(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) return findConn(head);
        }
        return nullptr;
    }
};
